File: week_1/day-2-prefix-sum-hashing/variations_que/prefix_2d.py
# prefix_2d.py

def buildPrefix2D(matrix):
    m, n = len(matrix), len(matrix[0])
    prefix = [[0]*(n+1) for _ in range(m+1)]

    for i in range(m):
        for j in range(n):
            prefix[i+1][j+1] = (
                matrix[i][j]
                + prefix[i][j+1]
                + prefix[i+1][j]
                - prefix[i][j]
            )
    return prefix

def query(prefix, r1, c1, r2, c2):
    return (
        prefix[r2+1][c2+1]
        - prefix[r1][c2+1]
        - prefix[r2+1][c1]
        + prefix[r1][c1]
    )

# Example Input
matrix = [
    [1, 2, 3],
    [4, 5, 6],
    [7, 8, 9]
]

prefix = buildPrefix2D(matrix)
print(query(prefix, 0, 0, 1, 1))  # Output: 12 (sum of top-left 2x2 submatrix)
print(query(prefix, 1, 1, 2, 2))  # Output: 28 (sum of bottom-right 2x2 submatrix)

File: week_1/day-2-prefix-sum-hashing/variations_que/prefix_2d.js
// prefix_2d.js

function buildPrefix2D(matrix) {
    const m = matrix.length, n = matrix[0].length;
    const prefix = Array.from({ length: m + 1 }, () => Array(n + 1).fill(0));

    for (let i = 0; i < m; i++) {
        for (let j = 0; j < n; j++) {
            prefix[i+1][j+1] =
                matrix[i][j]
                + prefix[i][j+1]
                + prefix[i+1][j]
                - prefix[i][j];
        }
    }
    return prefix;
}

function query(prefix, r1, c1, r2, c2) {
    return (
        prefix[r2+1][c2+1]
        - prefix[r1][c2+1]
        - prefix[r2+1][c1]
        + prefix[r1][c1]
    );
}

// Example Input
const matrix = [
    [1, 2, 3],
    [4, 5, 6],
    [7, 8, 9]
];

const prefix = buildPrefix2D(matrix);
console.log(query(prefix, 0, 0, 1, 1)); // Output: 12
console.log(query(prefix, 1, 1, 2, 2)); // Output: 28
